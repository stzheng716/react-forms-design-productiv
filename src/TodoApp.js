import React, { useState } from "react";
import { v4 as uuid } from "uuid";

import TopTodo from "./TopTodo";
import EditableTodoList from "./EditableTodoList";
import TodoForm from "./TodoForm";

/** App for managing a todo list.
 *
 * Props:
 * - initialTodos: possible array of [ todo, ... ]
 *
 * State:
 * - todos: array of [ todo, ... ]
 *
 * App -> TodoApp -> { TodoForm, EditableTodoList }
 */

function TodoApp({ initialTodos=[] }) {
  const [todos, setTodos] = useState(initialTodos)

  // const blankFormData = {title: "", description: "", priority: 1} //TODO: move to Form

  /** add a new todo to list */
  function create(newTodo) {
    setTodos(todos => [...todos, {...newTodo, id: uuid()}])
  }

  /** update a todo with updatedTodo */
  function update(updatedTodo) {
    setTodos(todos => todos.map(
      (todo) => todo.id === updatedTodo.id ? updatedTodo : todo
    ))
  }

  /** delete a todo by id */
  function remove(id) {
    console.log("todo id", id)
    setTodos(todos => todos.filter(
      (todo) => todo.id !== id
    ))
  }

  return (
    <main className="TodoApp">
      <div className="row">

        <div className="col-md-6">
          {todos.length > 0 &&
              <EditableTodoList
              todos={todos}
              update={update}
              remove={remove} />}
          {todos.length === 0 &&
              <span className="text-muted">You have no todos.</span>}
        </div>

        <div className="col-md-6">
          {todos.length > 0 &&
              <section className="mb-4">
                <h3>Top Todo</h3>
                <TopTodo todos={todos} />
              </section>}

          <section>
            <h3 className="mb-3">Add NÃ¼</h3>
            <TodoForm handleSave={create} />
          </section>
        </div>

      </div>
    </main>
  );
}

export default TodoApp;