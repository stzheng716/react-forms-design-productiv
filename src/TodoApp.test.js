import React from "react";
import { render, fireEvent, toContainHTML } from "@testing-library/react";
import TodoApp from "./TodoApp";

let container, debug, title, description, priority, topTodo;

beforeEach(() => {
    const renderResult = render(<TodoApp />);
    container = renderResult.container
    debug = renderResult.debug

    title = container.querySelector("#newTodo-title")
    description = container.querySelector("#newTodo-description")
    priority = container.querySelector("#newTodo-priority")

})

it("TodoApp renders without crashing", function () {
    render(<TodoApp />);
});

it("matches snapshot", function () {
    expect(container).toMatchSnapshot();
});

it("Test that todo(verify w/ title) isn't there, add it, verify it's there", function () {
    // Add a new todo
    expect(container).toContainHTML("You have no todos.")
    fireEvent.change(title, {target: { value: "test title" }});
    fireEvent.change(description, {target: { value: "test description" }});
    fireEvent.change(priority, {target: { value: 2 }});
    fireEvent.submit(container.querySelector(".NewTodoForm-addBtn"))

    // Verify new todo is added to TodoList
    expect(container).not.toContainHTML("You have no todos.");
    expect(container).toContainHTML("test title");
});

it("Test that todo(verify w/ title) IS there, delete, and verify absent", function () {
    // Add a new todo & verify added to TodoList
    expect(container).toContainHTML("You have no todos.")
    fireEvent.change(title, {target: { value: "test title" }});
    fireEvent.submit(container.querySelector(".NewTodoForm-addBtn")) // submit new todo
    expect(container).not.toContainHTML("You have no todos.")

    // Delete todo & verify TodoList is empty
    fireEvent.click(container.querySelector(".EditableTodo-delBtn"))
    expect(container).toContainHTML("You have no todos.")
});

it("Test that EDIT todo works", function () {
    // Add a new todo
    expect(container).toContainHTML("You have no todos.")
    fireEvent.change(title, {target: { value: "test title" }});
    fireEvent.change(description, {target: { value: "test description" }});
    fireEvent.change(priority, {target: { value: 2 }});
    fireEvent.submit(container.querySelector(".NewTodoForm-addBtn")) // added todo

    // Verify todo is added to TodoList
    expect(container).not.toContainHTML("You have no todos.");
    expect(container).toContainHTML("test title");

    // Edit todo
    fireEvent.click(container.querySelector(".EditableTodo-toggle")) // click edit btn
    expect(container).toContainHTML("<form"); // edit form appears
    title = container.querySelector("#newTodo-title");
    fireEvent.change(title, {target: { value: "edited title" }});
    fireEvent.submit(container.querySelector(".NewTodoForm-addBtn")) // save changes

    // Verify edits updated on TodoList
    expect(container).toContainHTML("edited title") // check for edited title
    expect(container).not.toContainHTML("test title");

});

it("Test TopTodo Added", function () {
    // Add a new todo
    expect(container).toContainHTML("You have no todos.")
    fireEvent.change(title, {target: { value: "test title" }});
    fireEvent.change(description, {target: { value: "test description" }});
    fireEvent.change(priority, {target: { value: 2 }});
    fireEvent.submit(container.querySelector(".NewTodoForm-addBtn")) // added todo

    // verify todo is added to TodoList
    expect(container).not.toContainHTML("You have no todos.");
    expect(container).toContainHTML("test title");

    // verify todo is added to topTodo component
    topTodo = container.querySelector("#topTodo");
    expect(topTodo).toContainHTML("test title");

});

it("Test TopTodo Deleted", function () {
    // Add a new todo
    expect(container).toContainHTML("You have no todos.")
    fireEvent.change(title, {target: { value: "test title" }});
    fireEvent.change(description, {target: { value: "test description" }});
    fireEvent.change(priority, {target: { value: 2 }});
    fireEvent.submit(container.querySelector(".NewTodoForm-addBtn")) // added todo

    // verify todo is added to TodoList & TopTodo
    expect(container).not.toContainHTML("You have no todos.");
    expect(container).toContainHTML("test title");
    topTodo = container.querySelector("#topTodo");
    expect(topTodo).toContainHTML("test title");

    // delete added todo
    fireEvent.click(container.querySelector(".EditableTodo-delBtn"))

    // verify todo is deleted from TodoList & TopTodo
    expect(container).toContainHTML("You have no todos.")
    expect(container).not.toContain(topTodo);

});

it("Test TopTodo Edited", function () {
    // Add a new todo
    expect(container).toContainHTML("You have no todos.")
    fireEvent.change(title, {target: { value: "test title" }});
    fireEvent.change(description, {target: { value: "test description" }});
    fireEvent.change(priority, {target: { value: 2 }});
    fireEvent.submit(container.querySelector(".NewTodoForm-addBtn")) // added todo

    // verify todo is added to TodoList & TopTodo
    expect(container).not.toContainHTML("You have no todos.");
    expect(container).toContainHTML("test title");
    topTodo = container.querySelector("#topTodo");
    expect(topTodo).toContainHTML("test title");

    // Edit todo from TodoList
    fireEvent.click(container.querySelector(".EditableTodo-toggle")) // click edit btn
    expect(container).toContainHTML("<form"); // edit form appears
    title = container.querySelector("#newTodo-title");
    fireEvent.change(title, {target: { value: "edited title" }});
    fireEvent.submit(container.querySelector(".NewTodoForm-addBtn")) // save changes

    // Verify edits are present in TodoList & TopTodo
    expect(container).toContainHTML("edited title"); // check for edited title
    expect(container).not.toContainHTML("test title");
    expect(topTodo).toContainHTML("edited title");

});

