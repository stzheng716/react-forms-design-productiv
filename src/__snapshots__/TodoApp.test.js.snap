// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <main
    class="TodoApp"
  >
    <div
      class="row"
    >
      <div
        class="col-md-6"
      >
        <span
          class="text-muted"
        >
          You have no todos.
        </span>
      </div>
      <div
        class="col-md-6"
      >
        <section>
          <h3
            class="mb-3"
          >
            Add Nü
          </h3>
          <form
            class="NewTodoForm"
          >
            <div
              class="mb-3"
            >
              <input
                aria-label="Title"
                class="form-control"
                id="newTodo-title"
                name="title"
                placeholder="Title"
                value=""
              />
            </div>
            <div
              class="mb-3"
            >
              <textarea
                aria-label="Description"
                class="form-control"
                id="newTodo-description"
                name="description"
                placeholder="Description"
              />
            </div>
            <div
              class="mb-3 d-flex justify-content-between"
            >
              <div
                class="w-75 d-flex justify-content-between"
              >
                <label
                  class="d-inline-flex"
                  for="newTodo-priority"
                >
                  Priority:  
                </label>
                <select
                  class="form-control form-control-sm d-inline-flex"
                  id="newTodo-priority"
                  name="priority"
                >
                  <option
                    value="1"
                  >
                    Ultra-Über
                  </option>
                  <option
                    value="2"
                  >
                    Über
                  </option>
                  <option
                    value="3"
                  >
                    Meh
                  </option>
                </select>
              </div>
              <button
                class="btn-primary rig btn btn-sm NewTodoForm-addBtn"
              >
                Gø!
              </button>
            </div>
          </form>
        </section>
      </div>
    </div>
  </main>
</div>
`;
